MP Config compliance:

ikonkere opened the pull request:
* Fixed array parsing from toString();
* Implemented property names iteration.

ikonkere committed:
MP Config compliance:

* Fixed array parsing from toString();
* Implemented property names iteration.

ikonkere commented:
Just to clarify: there was a bug with injecting array properties (essentially injecting an array of literally `"{a", " b}"` instead of `"a", "b"`, if you want i can create an issue for it.

urbim commented:
Correct me if I'm wrong, but if I understand this right, you are trying to do the following:

```yaml
myPets: {a,b}
```

and

```java
 @Inject @ConfigProperty(name="myPets") private String[] myArrayPets;
```

If so, this is incorrect according to Config MP specification, the curly braces are not supposed to be there, use this instead:

```yaml
myPets: a,b
```

For more information, see: https://download.eclipse.org/microprofile/microprofile-config-1.3/microprofile-config-spec.html#_array_converters

ikonkere commented:
Of course not. i'm trying to do YAML i.e.
```
myPets:
- a
- b
```
What you write isn't even YAML-compatible. I can attach a simple failng test-case if you want.

ikonkere commented:
I careated a test that fails injecting arrays and/or lists without my fix:
https://github.com/ikonkere/kumuluzee-config-mp/blob/mp-config-test/kumuluz-mp-config-test.zip

urbim commented:
Thanks for the explanation, I get it now.

Unfortunately, MP Config specification doesn't have any notion of hierarchy in configuration key as of yet, whereas our configuration framework supported this hierarchy from the start by choosing yaml file as the primary configuration source. The change you are proposing still breaks compliance with the MP Config specification.

I would recommend you fix the issue here: https://github.com/kumuluz/kumuluzee-config-mp/blob/14d23524a499a7dddc7dda08334d6b59e5123bb4/src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java#L61

First check if return value of the key is a list, then start collecting values while validating that all values are "simple" (e.g. no nested objects). Then join all values in a comma separated string and return that.

ikonkere commented:
>>The change you are proposing still breaks compliance with the MP Config specification.

Hmmm, can you point at what exactly it breaks in the spec? Not challenging or anything, just for my info. Also, shouldn't the TCK fail in such cases?

>>I would recommend you fix the issue here:

Will get right at it, thanks for comprehensive review!

urbim commented:
From the link above:

> The delimiter for the config value is ",". The escape character is "\".

This means that string `"{a,b}"` should be parsed to array with two strings: `"{a"` and `"b}"`.

TCK doesn't check every edge scenario (in your case presence of braces), that would be impossible. The specification is still the ground truth.

ikonkere commented:
>>in your case presence of braces

Braces are put in the string by Kumuluz, not by end-user code, specifically by this line: https://github.com/kumuluz/kumuluzee/blob/62f1ea9e6824b67c7b66391a62da165f944bf000/common/src/main/java/com/kumuluz/ee/configuration/sources/FileConfigurationSource.java#L154 (invoking toString() on array value). It has nothing to do with the spec itself, but rather with how Kumuluz translates parameter values from one place to another. That's why i have originally fixed it in ConfigImpl.

>> First check if return value of the key is a list

That's also why your suggestion above is impossible in `ConfigSourceAdapter` imho, because "[", "]" and "," are valid characters on pure String parameters and therefore can't be used to infer the parameter to be of array type.

urbim commented:
Yes, but user could easily specify a configuration value `"{a,b}"` which would then in your case be "fixed" to return wrong values `a` and `b` instead of the correct ones `{a` and `b}`. That's why your fix needs to be applied only if the braces come from invoking toString() on an array.

ikonkere commented:
Yes, you are right, i just realized that's the case as well. But still, it's impossible to do that in the `ConfigSourceAdapter` as it has no info on parameter type. Will think of something else, but for now i'm outta ideas besides fixing it in `FileConfigurationSource` in the first place .

urbim commented:
`configurationSource.getListSize(key)` will return `Optional.empty()` if value is not a list, similarly `getMapKeys(key)` will return `Optional.empty()` if value is not a nested object.

ikonkere committed:
Proper fix for array parsing

* Reverted ConfigImpl#convertList to previous version;
* Modified ConfigImpl#ARRAY_SEPARATOR_REGEX to take into account
trailing whitespace after ",";
* Added array check to ConfigSourceAdapter#getValue.

ikonkere commented:
Ok, so now i think i got it right indeed :).

ikonkere commented:
Damn, where are the tests that fail, i can't find those in the source code?

ikonkere committed:
Fixed failing TCK

* Reverted ConfigImpl#ARRAY_SEPARATOR_REGEX as it seems TCK relies on a
whitespace after "," in alot of tests for some reason.
* Added ConfigSourceAdapter#ARRAY_SEPARATOR_REGEX because i must reuse
it still to remove whitespaces.

ikonkere commented:
I can't reproduce those fails:
```
[ERROR] Failures: 
[ERROR]   CDIPlainInjectionTest>Arquillian.arquillianBeforeClass:109 » Deployment Except...
[ERROR]   CDIPropertyNameMatchingTest>Arquillian.arquillianBeforeClass:109 » Deployment ...
[ERROR]   ConfigProviderTest>Arquillian.run:164->testEnvironmentConfigSource:97 » StringIndexOutOfBounds
```
they pass for me. It seems there's something with #getListSize() on certain environments, can you check that those fail/pass for you?

ikonkere commented:
Besides, i noticed that the tests still run against KumuluzEE 3.0.0, should it be upgraded, or that's intentional?

urbim commented:
Yes, you should try updating to 3.5.0, we usually update dependencies as we need them or before release.

ikonkere committed:
Proper testing

* Incremented KumuluzEE version to current;
* Added array/list injection tests from config.yaml.

ikonkere commented:
Aha, that did the trick, so i guess we could continue discussing the change :)

urbim commented on src/main/java/com/kumuluz/ee/config/microprofile/ConfigImpl.java:
```diff
@@ -143,14 +148,28 @@ public ConfigImpl(List<ConfigSource> configSources, Map<Type, Converter> convert
+		try {
+			return arg0.getPropertyNames().stream();
+		}
+		catch (Exception e) {
```
Why is this catch-all required? It would probably be better if we don't catch anything and just forward the exception.

ikonkere replied:
It guards against specific ConfigSource implementations that i happened to use this for, but you're correct, i will remove it.

urbim commented on src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java:
```diff
@@ -55,10 +58,17 @@ public ConfigSourceAdapter(ConfigurationSource configurationSource) {
     @Override
     public String getValue(String s) {
-        return configurationSource.get(s).orElse(null);
+    	String val = configurationSource.get(s).orElse(null);
```
We cannot rely on the curly braces format here, that is specific of yaml configuration source. Env and system property configuration sources return Optional.empty() if get() for list is called.

Instead the way to go here is to check if the value is a list via getListSize and if it is iterating through values (while validating there are no nested objects) and collecting them to a comma separated string.

ikonkere replied:
>>We cannot rely on the curly braces format here, that is specific of yaml configuration source

Not sure i get what you mean: there're no curly braces anywhere and "[, ]" notion is the Java common string representation for arrays.

>>iterating through values

But it's a string, how can i iterate through values without parsing it first?

ikonkere replied:
>>iterating through values (while validating there are no nested objects) and collecting them to a comma separated string.

In fact there's no need to do any of that: the corresponding ConfigSource has done it already. The value you get from it is the leaf value of YAML tree which is either a primitive, a String or a collection of the above.

urbim replied:
Yes, but the value isn't necessary from a yaml source, it can be an environment variable, system property, etcd value or even something implemented by the user.

See here for an example: https://github.com/kumuluz/kumuluzee/blob/392d71b11dae349bf002f3c2b854943c3db5e062/core/src/main/java/com/kumuluz/ee/factories/EeConfigFactory.java#L122

ikonkere replied:
Ah, you mean that supplying arrays is only supported by `FileConfigurationSource` and not by other two that kumuluzee-common has (and the etcd one)? Indeed it's true, but imho `ConfigSourceAdapter` isn't the right place to implement that precisely because defining an array/list may well be source-specific and you will never get it right in a common adapter.

I think we must agree upon the contract that a `com.kumuluz.ee.configuration.ConfigurationSource` must deal with it within itself (i will modify the method JavaDoc to mention that) and provide an array string representation from #get(String) (as the file one does). Then we implement that in each source (which is architecturally sound and #getListSize(String) in each of the sources already does all of the array parsing work anyway) and my fix does just what it does now.

What do you think?

urbim replied:
Defining array isn't source-specific, it is well defined: https://github.com/kumuluz/kumuluzee/wiki/Configuration#configuration-parameter-keys

It has to be because the whole configuration framework is organized in a hierarchy and if you want to override a key in a higher priority source the keys have to be same.

ikonkere replied:
Ah, now i see what you mean. Then we will have to accept that for YAML sources we are doing same work twice, because the array that `FileConfigurationSource` has read from config.yaml has already been analysed, collected and processed.

I will do as you suggested, though I still think that the code snippet you refered to: https://github.com/kumuluz/kumuluzee/blob/392d71b11dae349bf002f3c2b854943c3db5e062/core/src/main/java/com/kumuluz/ee/factories/EeConfigFactory.java#L122 is not the right way this must be done; you must be able to get an array of values from a ConfigSource because the low-level contracts for arrays/lists are different (e.g. storing in the database or Zookeeper trees, or even YAML) and the end-user should not do all that stuff iterating through values.

urbim commented on src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java:
```diff
@@ -55,10 +58,17 @@ public ConfigSourceAdapter(ConfigurationSource configurationSource) {
-        return configurationSource.get(s).orElse(null);
+    	String val = configurationSource.get(s).orElse(null);
+    	
+    	if (configurationSource.getListSize(s).orElse(0) > 0) {
```
Also use Optional.isEmpty() / Optional.ifPresent() instead of .orElse(0).

urbim commented on src/test/java/com/kumuluz/ee/config/microprofile/tests/ArrayInjectionTest.java:
```diff
@@ -0,0 +1,65 @@
+package com.kumuluz.ee.config.microprofile.tests;
```
Add license and javadoc on class with at lease `@since` and `@author`.

urbim reviewed (changes requested):
Great to hear! I added some comments.

ikonkere committed:
Generic array handling

* removed exception handling form ConfigImpl#asStream;
* removed pattern replacing form ConfigSourceAdapter#getValue;
* implemented proper array building.

urbim commented on src/main/java/com/kumuluz/ee/config/microprofile/ConfigImpl.java:
```diff
@@ -143,14 +153,22 @@ public ConfigImpl(List<ConfigSource> configSources, Map<Type, Converter> convert
+	 * @return stream of property names, never null
+	 */
+	private Stream<String> asStream(ConfigSource arg0) {
+		return arg0.getPropertyNames().stream();
```
This can be inlined

ikonkere replied:
It might, but this way it can also be reused and it looks cleaner :)!

urbim replied:
Reused where? And no, it doesn't look cleaner.

urbim commented on src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java:
```diff
@@ -55,10 +57,19 @@ public ConfigSourceAdapter(ConfigurationSource configurationSource) {
     @Override
     public String getValue(String s) {
-        return configurationSource.get(s).orElse(null);
+    	String val = configurationSource.get(s).orElse(null);
```
If this is null return immediately. We don't want any null pointer exceptions in getListSize.

urbim commented on src/main/java/com/kumuluz/ee/config/microprofile/adapters/ConfigSourceAdapter.java:
```diff
@@ -93,4 +104,35 @@ private void buildPropertiesMap(Map<String, String> map, String prefix) {
+
+	        //array item is an object, so we just toString() it
+	        if (objectKeys.isPresent()) {
+	            Map<String, String> map = new HashMap<>();
```
If array item is an object just skip it. As said previously the Config MP specification does not support key hierarchy. If users need this functionality they will be better off using our configuration framework until proper hirearchy gets added to Config MP specification.

ikonkere replied:
>>If array item is an object just skip it.

Won't this break the consistency of the array? I think we should at least maintain the expected size and put nulls or blanks on the corresponding indexes.

urbim replied:
No, if we process these object in any way they would be inconsistent with the configuration supplied.

ikonkere replied:
I don't get what you just said, if you put `sb.append("")` instead of the code block you quoted, no processing would happen whatsoever.

urbim replied:
Exactly. That's what I meant by skipping object.

ikonkere committed:
Further changes as requested

* Removed object handling while building the array
* inlined property names streaming.

urbim commented:
LGTM! Thanks for your contribution!
